📖 Proposal for Adding Opentelemetry Tracing support (#5982)
kubernetes-sigs/cluster-api-provider-aws: docs/proposal/20260425-opentelemetry-tracing-support.md

- Goals  
- Non Goals
- Proposal  
- User Stories  
- Implementation Details  
- Risks and Mitigations  
- Alternatives  
- Implementation Plan  
- Resources  

## Summary
With the increasing adoption of CAPA we want to improve the supportability of CAPA especially in production. CAPA currently lacks OpenTelemetry support for tracing in controllers. This proposal suggests adding OpenTelemetry support to its controllers enabling tracing for better observability.

## Motivation

### Problem Statement
With the increasing adoption of CAPA we want to improve the supportability of CAPA especially in production. CAPA currently lacks tracing support for its controllers making it difficult to track failures in controllers during reconciliation. Adding support for tracing would also benefit maintainers by enabling them to use tracing while investigating issues, flows etc.

## Goals
- OpenTelemetry SDK initialization and configuration in CAPA  
- Root spans at reconciliation boundaries for core CAPA controllers  
- Enable tracing for debugging and performance analysis  
- Add configurable sampling via flags  
- Enable trace collection locally when using tilt

## Non-Goals

> - Enable/disable tracing at runtime without restarting  

Reviewer: Is this something we want to add in the future? I can imagine this would be useful, but not necessary, in production. Users can always modify the CAPA Deployment.

Reviewer 2: Definitely a non-goal as we don't allow enabling things without a restart for other functionality.

> - Tail-based sampling  

Reviewer: Please explain why we won't do tail-based sampling. Not sure where... maybe in the risks & mitigations section?

Author: done

**Proposed location:**

    pkg/otel

**Responsibilities:**
- Initialize tracing  
- Provide reusable helpers  
- Configure exporters/resources  

This approach ensures maintainability and avoids repetitive instrumentation code in each controller.

### Trace Export

Tracing data will be exported using the OpenTelemetry Protocol (OTLP) exporter.

CAPA will configure a single OTLP exporter over grpc (`otlptracegrpc`) within the controller process to send trace data to a user-defined endpoint. This endpoint will be configurable via controller flags.

The configured endpoint may be:
- An OpenTelemetry Collector (recommended for production), which can process, batch, and route traces to one or more backends.
- A compatible backend (e.g., Grafana Tempo or Jaeger) that supports OTLP ingestion.

### High-Level Architecture

    CAPA Controllers
          ↓
    OpenTelemetry Go SDK
          ↓
    OTLP Exporter
          ↓
    OpenTelemetry Collector
          ↓
    Jaeger / Grafana Tempo / Other Backends

### Sampling Strategy

To control the volume of traces and reduce performance overhead, sampling will be applied.

> Sampling will be implemented using a ParentBased TraceIDRatioBased sampler, with the sampling ratio configurable via flags. It balances risk of OOM (which is a very possible risk in case of multiple large traces), operational visibility and system performance.

Reviewer: nit

```suggestion
Sampling will be implemented using a ParentBased TraceIDRatioBased sampler, with the sampling ratio configurable via flags. It balances the risk of exceeding memory limits, which is possible in case of multiple large traces, operational visibility, and system performance.
```

Sampling will be:
- Configurable via controller flags
- Applied at the trace level

Example configuration:
- Low sampling rate in production to minimize overhead
- Higher sampling rate during debugging or incident analysis

This ensures a balance between observability and system performance.
 
### Tracing AWS API Calls

AWS API calls can be traced using OpenTelemetry instrumentation for AWS SDK (`otelaws` middleware). This enables automatic creation of spans for all outgoing AWS API requests. Each AWS API span will be a child of the reconciliation span. it will be optional to trace AWS API call and can be enabled by flags.

> However, this approach may generate excessive spans and introduce noise. Instead we can also create spans for important AWS API Calls For Ex- `CreateInstance`

Reviewer: ```suggestion
However, this approach may generate excessive spans and introduce noise. Instead we can also create spans for important AWS API Calls, e.g.,  `CreateInstance`.
```

To address this we can manually trace instrument important AWS operations such as CreateInstance or DeleteLoadBalancer etc.

> ### Functions to Trace

Author: here we need to discuss the need to trace the services as well.

Reviewer: Yes. I think you mean the CAPA "services" abstraction. They are the layer that interacts with the AWS APIs, so we definitely want to trace them.

> To address this we can manually trace instrument important AWS operations such as CreateInstance or DeleteLoadBalancer etc.

Reviewer: Is this line repeating information in the previous paragraph?

Author: Yes, caught by over-explanation!

### Additional Trace Data
We can attach some useful data to traces which will give us meaningful review of the traces

**Span Attributes:**

- **Kubernetes metadata**
  - `cluster.name`
  - `namespace`
  - `resource.name`
  - `resource.kind`

- **Reconciliation metadata**
  - `reconcile.id`
  - `reconcile.result`
  - `reconcile.duration`

- **AWS metadata**
  - `aws.resource.id`
  - `aws.region`

### Combined Event Recorder

We can explore a unified approach where Kubernetes events are recorded using the standard event recorder and also attached to OpenTelemetry spans.

This would allow better correlation between Kubernetes events and traces. Further investigation is required to determine the best implementation approach.

### Handling logs in relation to spans

We will explore correlating logs with traces by including trace and span IDs in structured logs. This will allow easier navigation between logs and traces.

Further investigation is required to finalize this approach.

### Exposed Configurations

| Flag                    | Description               |
|-------------------------|---------------------------|

> | `--otel-enabled`        | Enable/disable tracing    |

Reviewer 2: I don't think we need a separate enabled flag. If `--otel-endpoint` is supplied and is not an empty string then its enabled :)

> | `--otel-endpoint`       | OTLP endpoint (collector or backend)   |

Reviewer 2: ```suggestion
| `--tracing-endpoint`       | Trace endpoint (collector or backend)   |
```

> | `--otel-sampling-rate`  | Sampling ratio            |

Reviewer 2: ```suggestion
| `--tracing-sampling-rate`  | Sampling ratio            |
```

> | `--otelaws-enabled`     | Enable/disable aws api calls |

Reviewer 2: ```suggestion
| `--tracing-aws-services`     | Enable/disable aws api calls |
```
